model/pod/pod.go

```diff
@@ -156,20 +184,100 @@ func (p *Pod) Remove() error {
 	)
 }
 
-// UpdateStatus updates the pod status.
+// UpdateStatus updates the pod status. If the new status is identical to the
+// current one, this is a no-op.
 func (p *Pod) UpdateStatus(s Status) error {
+	if p.Status == s {
+		return nil
+	}
+
 	byIDAndStatus := ByID(p.ID)
 	byIDAndStatus[StatusKey] = p.Status
 
+	updated := utility.BSONTime(time.Now())
+	setFields := bson.M{StatusKey: s}
+	switch s {
+	case StatusInitializing:
+		setFields[bsonutil.GetDottedKeyName(TimeInfoKey, TimeInfoInitializingKey)] = updated
+	case StatusStarting:
+		setFields[bsonutil.GetDottedKeyName(TimeInfoKey, TimeInfoStartingKey)] = updated
+	}
+
 	if err := UpdateOne(byIDAndStatus, bson.M{
-		"$set": bson.M{StatusKey: s},
+		"$set": setFields,
 	}); err != nil {
 		return err
 	}
 
 	event.LogPodStatusChanged(p.ID, string(p.Status), string(s))
 
 	p.Status = s
+	switch s {
+	case StatusInitializing:
+		p.TimeInfo.Initializing = updated
+	case StatusStarting:
+		p.TimeInfo.Starting = updated
+	}
 
 	return nil
 }
+
+// UpdateResources updates the pod resources. If the new status is identical to the current one, this is a no-op.
+func (p *Pod) UpdateResources(info ResourceInfo) error {
+	if p.Resources.Cluster == info.Cluster && p.Resources.ExternalID == info.ExternalID && p.Resources.DefinitionID == info.DefinitionID && p.checkSecretIDs(p.Resources.SecretIDs, info.SecretIDs) {
+		return nil
+	}
+
+	if info.Cluster == "" {
+		info.Cluster = p.Resources.Cluster
+	}
+
+	if info.ExternalID == "" {
+		info.ExternalID = p.Resources.ExternalID
+	}
+
+	if info.DefinitionID == "" {
+		info.DefinitionID = p.Resources.DefinitionID
+	}
+
+	if len(info.SecretIDs) == 0 {
+		info.SecretIDs = p.Resources.SecretIDs
+	}
+
+	setFields := bson.M{
+		ResourcesKey: info,
+	}
+
+	if err := UpdateOne(ByID(p.ID), bson.M{
+		"$set": setFields,
+	}); err != nil {
+		return err
+	}
+
+	p.Resources = info
+
+	return nil
+}
+
+// checkSecretIDs checks to see if the two given slices have the same IDs.
+func (p *Pod) checkSecretIDs(oldIDs, newIDs []string) bool {
```

Comment:
I assume you're going to remove this, but you can achieve the same thing using `utility.SliceSliceSymmetricDifference`.